shouldMaintainOrderingWithAsyncMessages can cause unrelated async messages to be reordered.

mattwoodrow opened the pull request:
#### a8242ccbe1fe4db7f6c673018283a06fb2465173
<pre>
shouldMaintainOrderingWithAsyncMessages can cause unrelated async messages to be reordered.
<a href="https://bugs.webkit.org/show_bug.cgi?id=301678">https://bugs.webkit.org/show_bug.cgi?id=301678</a>
&lt;<a href="https://rdar.apple.com/problem/163697393">rdar://problem/163697393</a>&gt;

Reviewed by Kimmo Kinnunen.

We receive messages A, Sync (with the the
shouldMaintainOrderingWithAsyncMessages flag), B.

When A arrives, we put a callback for dispatchIncomingMessages into the
destination threads queue, and put A into m_incomingMessages.

When Sync arrives, it pulls A into the
m_messagesToDispatchWhileWaitingForSyncReply queue, along with Sync  and put a
callback for dispatchSyncStateMessages into the dest thread queue.

When B arrives, it gets put into m_incomingMessages.

Then on the dest thread, dispatchIncomingMessages runs, which dispatches B.
Sometime later on dispatchSyncStateMessages runs, which dispatches A,Sync.

Add a new HashSet that tracks when we have the
shouldMaintainOrderingWithAsyncMessages, and add following messages to the sync
ordering queue as well (up until the sync message gets dispatched).

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::processIncomingMessage):
(IPC::Connection::SyncMessageState::dispatchMessagesAndResetDidScheduleDispatchMessagesForConnection):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: <a href="https://commits.webkit.org/302457@main">https://commits.webkit.org/302457@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/672d3e2e666c78059eb4c129fcbb05dc0f8bc250

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129146 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1405 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39982 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136524 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80530 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6f4dcc0b-838e-459c-ab1f-c641614adf36) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1337 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1282 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98337 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/66209 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132093 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1042 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115688 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78985 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b1d52465-f4fa-4469-8639-375b90b17b61) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/964 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33805 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79804 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109411 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34301 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139001 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1198 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1161 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106871 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1250 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112024 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106706 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/985 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30548 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53773 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20166 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1271 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64623 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1097 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1142 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1195 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
```diff
@@ -582,6 +582,150 @@ TEST_P(ConnectionRunLoopTest, RunLoopSendAsync)
     localReferenceBarrier();
 }
 
+// Test for ensuring that async messages are not reordered when sync message is sent with SendSyncOption::MaintainOrderingWithAsyncMessages.
```
Could you also add the test variants that I sent?
The trouble with your added variants is that they do not test the primary function of MaintainOrderingWithAsyncMessages, e.g. that A, Sync, B arrives as A, Sync, B.
Otherwise do not know that  MaintainOrderingWithAsyncMessages actually works as it intends.

mattwoodrow committed:
shouldMaintainOrderingWithAsyncMessages can cause unrelated async messages to be reordered.
https://bugs.webkit.org/show_bug.cgi?id=301678
<rdar://problem/163697393>

Reviewed by Kimmo Kinnunen.

We receive messages A, Sync (with the the
shouldMaintainOrderingWithAsyncMessages flag), B.

When A arrives, we put a callback for dispatchIncomingMessages into the
destination threads queue, and put A into m_incomingMessages.

When Sync arrives, it pulls A into the
m_messagesToDispatchWhileWaitingForSyncReply queue, along with Sync  and put a
callback for dispatchSyncStateMessages into the dest thread queue.

When B arrives, it gets put into m_incomingMessages.

Then on the dest thread, dispatchIncomingMessages runs, which dispatches B.
Sometime later on dispatchSyncStateMessages runs, which dispatches A,Sync.

Add a new HashSet that tracks when we have the
shouldMaintainOrderingWithAsyncMessages, and add following messages to the sync
ordering queue as well (up until the sync message gets dispatched).

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::SyncMessageState::processIncomingMessage):
(IPC::Connection::SyncMessageState::dispatchMessagesAndResetDidScheduleDispatchMessagesForConnection):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::TEST_P):

Canonical link: https://commits.webkit.org/302457@main
